src/AstQualifiedName.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2015, Oracle and/or its affiliates. All rights reserved
+ * Licensed under the Universal Permissive License v 1.0 as shown at:
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+/************************************************************************
+ *
+ * @file AstQualifiedName.h
+ *
+ * Defines a class for qualified names so that components can be accessed.
+ * Qualified names are used for types and relations.
+ *
+ ***********************************************************************/
+
+#pragma once
+
+#include "Util.h"
+
+#include <algorithm>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace souffle {
+
+/**
+ * Qualified Name class
+ */
+class AstQualifiedName {
+public:
+    AstQualifiedName() : qualifiers() {}
+    AstQualifiedName(const std::string& name) : qualifiers({name}) {}
+    AstQualifiedName(const char* name) : AstQualifiedName(std::string(name)) {}
+    AstQualifiedName(const std::vector<std::string> qualifiers) : qualifiers(qualifiers) {}
+    AstQualifiedName(const AstQualifiedName&) = default;
+    AstQualifiedName(AstQualifiedName&&) = default;
+    AstQualifiedName& operator=(const AstQualifiedName&) = default;
+    AstQualifiedName& operator=(AstQualifiedName&&) = default;
+
+    /** append qualifiers */
+    void append(const std::string& name) {
+        qualifiers.push_back(name);
+    }
+
+    /** prepend qualifiers */
+    void prepend(const std::string& name) {
+        qualifiers.insert(qualifiers.begin(), name);
+    }
+
+    /** check for emptiness */
+    bool empty() const {
+        return qualifiers.empty();
+    }
+
+    /** get qualifiers */
+    const std::vector<std::string>& getQualifiers() const {
+        return qualifiers;
+    }
+
+    /** convert to a string separated by fullstop */
+    std::string toString() const {
+        std::stringstream ss;
+        ss << join(qualifiers, ".");
+        return ss.str();
+    }
+
+    bool operator==(const AstQualifiedName& other) const {
+        return qualifiers == other.qualifiers;
+    }
+
+    bool operator!=(const AstQualifiedName& other) const {
+        return !(*this == other);
+    }
+
+    bool operator<(const AstQualifiedName& other) const {
+        return std::lexicographical_compare(
```

Comment:
I think this is equivalent to `return qualifiers < other.qualifiers;`, which would be simpler and not require the algorithm include

Reply:
No - unfortunately, this gives a different semantics and we fail to execute benchmarks such as name_clash.dl.